### PR TITLE
sql: tighten parsing on GRANT and support multiple targets

### LINKED ISSUE #4093: SQL: `GRANT ALL ON TABLE testdb.* TO readwrite` crashes server
Database started in dev mode:
```
./cockroach start --dev
```

Small script to connect and crash db node:
```
package main

import (
	"database/sql"

	_ "github.com/lib/pq"
)

func main() {
	db, _ := sql.Open("postgres", "postgres://root@localhost:15432?sslmode=disable")
	db.Exec(`CREATE DATABASE IF NOT EXISTS testdb`)
	db.Exec(`GRANT ALL ON TABLE testdb.* TO readwrite`)
}
```

DB node stack trace (top of it):
```
panic: interface conversion: parser.IndirectionElem is parser.StarIndirection, not parser.NameIndirection

goroutine 147 [running]:
github.com/cockroachdb/cockroach/sql/parser.(*QualifiedName).Table(0xc8201408a0, 0x0, 0x0)
	github.com/cockroachdb/cockroach/sql/parser/expr.go:436 +0x1d7
github.com/cockroachdb/cockroach/sql.(*planner).getTableDesc(0xc820197000, 0xc8201408a0, 0xc82017c360, 0x1)
	github.com/cockroachdb/cockroach/sql/table.go:192 +0x14d
github.com/cockroachdb/cockroach/sql.(*planner).getDescriptorFromTargetList(0xc820197000, 0x0, 0x0, 0x0, 0xc82049e188, 0x1, 0x1, 0x0, 0x0, 0x3200000032)
	github.com/cockroachdb/cockroach/sql/descriptor.go:183 +0x29b
github.com/cockroachdb/cockroach/sql.(*planner).changePrivileges(0xc820197000, 0x0, 0x0, 0x0, 0xc82049e188, 0x1, 0x1, 0xc820248cf0, 0x1, 0x1, ...)
	github.com/cockroachdb/cockroach/sql/grant.go:26 +0x93
github.com/cockroachdb/cockroach/sql.(*planner).Grant(0xc820197000, 0xc820455500, 0x0, 0x0, 0x1)
	github.com/cockroachdb/cockroach/sql/grant.go:74 +0x90
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc820197000, 0x6e64df0, 0xc820455500, 0x145601, 0x0, 0x0, 0x145649)
	github.com/cockroachdb/cockroach/sql/plan.go:111 +0x8ad
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func1(0xece43123f, 0x32a5a22d, 0x59aafa0, 0xc820140801, 0x28)
	github.com/cockroachdb/cockroach/sql/executor.go:403 +0xd0
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func2(0xc8200ad540, 0x1)
	github.com/cockroachdb/cockroach/sql/executor.go:466 +0x121
github.com/cockroachdb/cockroach/client.(*Txn).exec(0xc8200ad540, 0xc82065e3b8, 0x14)
	github.com/cockroachdb/cockroach/client/txn.go:405 +0x2b2
github.com/cockroachdb/cockroach/client.(*DB).Txn(0xc8202ae580, 0xc82065e3b8, 0x6e64d60)
	github.com/cockroachdb/cockroach/client/db.go:463 +0x15b
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8202ae580, 0x6e64df0, 0xc820455500, 0xc820197000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cockroachdb/cockroach/sql/executor.go:463 +0x600
github.com/cockroachdb/cockroach/sql.(*Executor).execStmts(0xc8202ae580, 0xc82067e025, 0x3b, 0xc820197000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/cockroachdb/cockroach/sql/executor.go:318 +0x33d
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8202ae580, 0xc8204d601e, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/cockroachdb/cockroach/sql/executor.go:266 +0x5e5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82007e000, 0xc82067e025, 0x3b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x69ac101, ...)
	github.com/cockroachdb/cockroach/sql/pgwire/v3.go:550 +0x1fb
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc82007e000, 0xc82007e038, 0x0, 0x0)
	github.com/cockroachdb/cockroach/sql/pgwire/v3.go:274 +0xc5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82007e000, 0x0, 0x0, 0x0)
	github.com/cockroachdb/cockroach/sql/pgwire/v3.go:234 +0xf21
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serveConn(0xc8202a41b0, 0x7c001b8, 0xc82049e068, 0x0, 0x0)
	github.com/cockroachdb/cockroach/sql/pgwire/server.go:200 +0x9b5
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve.func1(0xc8202a41b0, 0x7c001b8, 0xc82049e068)
	github.com/cockroachdb/cockroach/sql/pgwire/server.go:109 +0x90
created by github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve
	github.com/cockroachdb/cockroach/sql/pgwire/server.go:114 +0x20a
```

### PR DESCRIPTION
**Pull request by andreimatei**
- restrict grammar only to valid targets for GRANT and REVOKE
- allow multiple targets for GRANT and REVOKE
- allow wildcards for GRANT, REVOKE and SHOW GRANTS.

Fixes #4093

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4158)

### CODE REVIEW

File: sql/testdata/grant_table [hunk 1]
```diff
@@ -4,6 +4,9 @@ CREATE DATABASE a
 statement ok
 GRANT ALL ON DATABASE a TO readwrite
 
+statement ok
```
Comment (vivekmenezes) [inline sql/testdata/grant_table]: Are you repeating this on purpose? Perhaps you want to add a comment on why.

Comment (andreimatei) [inline sql/testdata/grant_table]: Oops, it was not on purpose.

**Comment by petermattis** [comment]
LGTM

I've noticed that the postgres grammar sometimes does not outlaw invalid syntax, but leaves discovering such invalid syntax to the semantic analysis phase. I think we might be starting down a difficult path if we want the grammar to exactly parse the syntax we support.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/testdata/grant_table, line 120 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jBUaU-WCLYOaF7CKR:-K9jBUaVyvAwL9fxG9xH:2012839769) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f18340b726438755ee3352181bfc3bc5487dd2/sql/testdata/grant_table#L120)):*
nit: we usually put spaces around our `=`.

---

*[sql/testdata/grant_table, line 198 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jB_fKawOpFQ2Iik7F:-K9jB_fKawOpFQ2Iik7G:61039115) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f18340b726438755ee3352181bfc3bc5487dd2/sql/testdata/grant_table#L198)):*
Nice tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-:-K9jCrm_GahrWkhKZ6gO:-825862389)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/descriptor.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jDanxCmGJCjlZYACo:-K9jDanxCmGJCjlZYACp:-358093876) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L167)):*
wanna allocate this after the below condition

---

*[sql/descriptor.go, line 181 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jDpeH21-nRelegwqS:-K9jDpeH21-nRelegwqT:678021867) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L181)):*
same

---

*[sql/descriptor.go, line 218 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jElMRYFHlI5XBkJQ9:-K9jElMRYFHlI5XBkJQA:-200555088) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L218)):*
can this even happen or is this a panic condition?

---

*[sql/testdata/grant_table, line 136 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jAuLHaicCLjJdswdW:-K9jAuLHaicCLjJdswdX:-2081615179) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/testdata/grant_table#L136)):*
how about using names like b and c given that we already used "a"

---

*[sql/testdata/grant_table, line 190 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jA41t2SMlKf_VQhEf:-K9jA41t2SMlKf_VQhEg:-1433061578) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/testdata/grant_table#L190)):*
Your choice of Millie Vanilli makes me feel young.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158)*

**Comment by vivekmenezes** [comment]
LGTM

---

Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-:-K9jEyIts-OWoIcPr8Rb:676834233)*

**Comment by andreimatei** [comment]
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/descriptor.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jDanxCmGJCjlZYACo:-K9jKUfhKnhdUYoPZmee:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L167)):*
Done.

---

*[sql/descriptor.go, line 181 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jDpeH21-nRelegwqS:-K9jKV3M0Nwuv6uyrkqT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L181)):*
Done.

---

*[sql/descriptor.go, line 218 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jElMRYFHlI5XBkJQ9:-K9jLO4lh6fTo0BVCMSd:-122708540) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/descriptor.go#L218)):*
it would be a bug.
The way I understand it, `util.Errorf` already signifies a bug, passing this up is strictly better than a panic.

---

*[sql/testdata/grant_table, line 120 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jBUaU-WCLYOaF7CKR:-K9jLshAznhp0GKubuz7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9f18340b726438755ee3352181bfc3bc5487dd2/sql/testdata/grant_table#L120)):*
Done.

---

*[sql/testdata/grant_table, line 136 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jAuLHaicCLjJdswdW:-K9jMmcrfyRs2nH5-zGe:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/testdata/grant_table#L136)):*
Done.

---

*[sql/testdata/grant_table, line 190 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jA41t2SMlKf_VQhEf:-K9jLz0LvFlzTI3MtMLM:-1415671525) ([raw file](https://github.com/cockroachdb/cockroach/blob/17f37f3663619e2cb04d75abf7e783ab30b6b7c7/sql/testdata/grant_table#L190)):*
my policy with movies, for example, is that everything after 1985 is brand new.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158)*

**Comment by tamird** [comment]
Reviewed 5 of 7 files at r1, 1 of 1 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/testdata/grant_table, line 149 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jQ_8CgTlh_X3pB0DM:-K9jQ_8CgTlh_X3pB0DN:-1242518158) ([raw file](https://github.com/cockroachdb/cockroach/blob/70955d79186ea8f35cd6e6f7018287796938bd72/sql/testdata/grant_table#L149)):*
spaces

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158)*

**Comment by andreimatei** [comment]
@petermattis yeah... I was initially curious what it'd take to change the grammar, and then it was done so I left it. But I wouldn't necessarily advise this practice everywhere.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/testdata/grant_table, line 149 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-K9jQ_8CgTlh_X3pB0DM:-K9mT2niaNEe9bCVO1Ny:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/70955d79186ea8f35cd6e6f7018287796938bd72/sql/testdata/grant_table#L149)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4158#-:-K9mTE0cpH2oa_q3qqbw:201665611)*

**Pull request merged**